src/metrics/rules/active_ruleset.go

```diff
@@ -402,26 +427,36 @@ func (as *activeRuleSet) toRollupResults(
 		}
 		if rollupID == nil {
 			// The applied pipeline applies to the incoming ID.
-			pipelines = append(pipelines, newPipeline)
+			matchResult.forExistingID[feIdx].Pipelines =
+				append(matchResult.forExistingID[feIdx].Pipelines, newPipeline)
 		} else {
 			if len(tags[idx]) > 0 {
 				newPipeline.Tags = tags[idx]
 			}
 			// The applied pipeline applies to a new rollup ID.
-			matchResults := ruleMatchResults{
-				cutoverNanos: cutoverNanos,
-				pipelines:    []metadata.PipelineMetadata{newPipeline},
+			newRollupIDResult := IDWithMetadatas{
+				ID: rollupID,
+				Metadatas: metadata.StagedMetadatas{
+					{
+						CutoverNanos: cutoverNanos,
+						Metadata: metadata.Metadata{
+							Pipelines: []metadata.PipelineMetadata{newPipeline},
+						},
+					},
+				},
 			}
-			newRollupIDResult := idWithMatchResults{id: rollupID, matchResults: matchResults}
 			newRollupIDResults = append(newRollupIDResults, newRollupIDResult)
 		}
 	}
+	if matchResult.forExistingID[feIdx].CutoverNanos < cutoverNanos {
+		matchResult.forExistingID[feIdx].CutoverNanos = cutoverNanos
+	}
+	matchResult.keepOriginal = keepOriginal
 
-	return rollupResults{
-		forExistingID:   ruleMatchResults{cutoverNanos: cutoverNanos, pipelines: pipelines},
-		forNewRollupIDs: newRollupIDResults,
-		keepOriginal:    keepOriginal,
-	}, multiErr.FinalError()
+	sort.Sort(IDWithMetadatasByIDAsc(newRollupIDResults))
+	mergeResultsForNewRollupIDs(matchResult, newRollupIDResults, timeNanos)
```

Comment:
Now, we merge in `newRollupIDs` right away as opposed to waiting until we get here: https://github.com/m3db/m3/pull/4019/files#diff-2d6d0ddef3f98e8ba7f8d8a15c0367b893bd32dec1fcddf062c3c9034eca966bL122